ingest/usecase/plugins/orderbook/claimbot/plugin.go

```diff
@@ -0,0 +1,160 @@
+	}
+
+	return &claimbot{
+		config:     config,
+		atomicBool: atomic.Bool{},
+	}, nil
+}
+
+// ProcessEndBlock implements domain.EndBlockProcessPlugin.
+// This method is called at the end of each block to process and claim eligible orderbook orders.
+// ProcessEndBlock implements domain.EndBlockProcessPlugin.
+func (o *claimbot) ProcessEndBlock(ctx context.Context, blockHeight uint64, metadata domain.BlockPoolMetadata) error {
+	ctx, span := tracer.Start(ctx, "orderbooktFillerIngestPlugin.ProcessEndBlock")
+	defer span.End()
+
+	// For simplicity, we allow only one block to be processed at a time.
+	// This may be relaxed in the future.
+	if !o.atomicBool.CompareAndSwap(false, true) {
+		o.config.Logger.Info("already in progress", zap.Uint64("block_height", blockHeight))
+		return nil
+	}
+	defer o.atomicBool.Store(false)
+
+	orderbooks, err := getOrderbooks(o.config.PoolsUseCase, blockHeight, metadata)
+	if err != nil {
+		return err
+	}
+
+	// retrieve claimable orders for the orderbooks
+	orders := processOrderbooksAndGetClaimableOrders(
+		ctx,
+		fillThreshold,
+		orderbooks,
+		o.config.OrderbookRepository,
+		o.config.OrderBookClient,
+		o.config.OrderbookUsecase,
+		o.config.Logger,
+	)
+
+	for _, orderbook := range orders {
+		if orderbook.Err != nil {
+			o.config.Logger.Warn(
+				"failed to retrieve claimable orders",
+				zap.String("contract_address", orderbook.Orderbook.ContractAddress),
+				zap.Error(orderbook.Err),
+			)
+			continue
+		}
+
+		if err := o.processOrderbookOrders(ctx, orderbook.Orderbook, orderbook.Orders); err != nil {
+			o.config.Logger.Warn(
+				"failed to process orderbook orders",
+				zap.String("contract_address", orderbook.Orderbook.ContractAddress),
+				zap.Error(err),
+			)
+		}
+	}
+
+	o.config.Logger.Info("processed end block", zap.Uint64("block_height", blockHeight))
+
+	return nil
+}
+
+// processOrderbookOrders processes a batch of claimable orders.
+func (o *claimbot) processOrderbookOrders(ctx context.Context, orderbook domain.CanonicalOrderBooksResult, orders orderbookdomain.Orders) error {
+	if len(orders) == 0 {
+		return fmt.Errorf("no claimable orders found for orderbook %s, nothing to process", orderbook.ContractAddress)
+	}
+
+	for _, chunk := range slices.Split(orders, maxBatchOfClaimableOrders) {
+		if len(chunk) == 0 {
+			continue
+		}
+
+		txres, err := sendBatchClaimTx(
+			ctx,
+			o.config.Keyring,
+			o.config.AccountQueryClient,
+			o.config.TxfeesClient,
+			o.config.GasCalculator,
+			o.config.TxServiceClient,
+			o.config.ChainID,
+			orderbook.ContractAddress,
+			chunk,
+		)
+
+		if err != nil || (txres != nil && txres.Code != 0) {
+			o.config.Logger.Info("failed sending tx",
+				zap.String("orderbook contract address", orderbook.ContractAddress),
+				zap.Any("orders", chunk),
+				zap.Any("tx result", txres),
+				zap.Error(err),
+			)
+		}
+
+		// Wait for block inclusion with buffer to avoid sequence mismatch
+		time.Sleep(blockInclusionWaitTime)
```

Comment:
Consider querying the sequence number at the beginning of the block and incrementing it in-memory. This should be fine since we have a lock on block processing. That is, if block X is being processed, block X+1 processing cannot start.

This would drastically speed up processing all orders within a block

Reply:
Addressed that [here](https://github.com/osmosis-labs/sqs/pull/524/files#diff-9c91025a0f9e383b452c4f91a803200bdeb323f4511b300b44e5ab6bdf19c628R204) :+1: